src/Cli/dotnet/Commands/Run/VirtualProjectBuildingCommand.cs

```diff
@@ -142,28 +162,73 @@ public override int Execute()
         var consoleLogger = TerminalLogger.CreateTerminalOrConsoleLogger(MSBuildArgs.OtherMSBuildArgs.ToArray());
         var binaryLogger = GetBinaryLogger(MSBuildArgs.OtherMSBuildArgs);
 
-        RunFileBuildCacheEntry? cacheEntry = null;
+        CacheInfo? cache = null;
 
         if (!NoBuild)
         {
             if (NoCache)
             {
-                cacheEntry = ComputeCacheEntry(out _);
+                PrepareProjectInstance();
+
+                cache = ComputeCacheEntry();
+                LastBuildLevel = BuildLevel.All;
             }
-            else if (!NeedsToBuild(out cacheEntry))
+            else
             {
-                if (binaryLogger is not null)
+                PrepareProjectInstance();
+
+                var buildLevel = GetBuildLevel(out cache);
+                LastBuildLevel = buildLevel;
+
+                if (buildLevel is BuildLevel.None)
                 {
-                    Reporter.Output.WriteLine(CliCommandStrings.NoBinaryLogBecauseUpToDate.Yellow());
+                    if (binaryLogger is not null)
+                    {
+                        Reporter.Output.WriteLine(CliCommandStrings.NoBinaryLogBecauseUpToDate.Yellow());
+                    }
+
+                    return 0;
                 }
 
-                PrepareProjectInstance();
+                if (buildLevel is BuildLevel.Csc)
+                {
+                    if (binaryLogger is not null)
+                    {
+                        Reporter.Output.WriteLine(CliCommandStrings.NoBinaryLogBecauseRunningJustCsc.Yellow());
+                    }
+
+                    MarkBuildStart();
 
-                return 0;
+                    // Execute CSC.
+                    int result = new CSharpCompilerCommand
+                    {
+                        EntryPointFileFullPath = EntryPointFileFullPath,
+                        ArtifactsPath = ArtifactsPath,
+                        // We can reuse auxiliary files if we previously built using csc.
+                        CanReuseAuxiliaryFiles = cache.CanReuseAuxiliaryFiles && cache.PreviousEntry?.BuildLevel is BuildLevel.Csc,
+                    }
+                    .Execute(out bool fallbackToNormalBuild);
+
+                    if (result == 0)
+                    {
+                        Debug.Assert(!fallbackToNormalBuild);
+                        MarkBuildSuccess(cache);
+                    }
+
+                    if (!fallbackToNormalBuild)
+                    {
+                        return result;
+                    }
```

Comment:
nit: Consider a factoring like the following which in my opinion makes it more "structurally" obvious that we are not going to fallback after marking a build a success.

```suggestion
                    if (!fallbackToNormalBuild)
                    {
                        if (result == 0)
                        {
                            MarkBuildSuccess(cache);
                        }

                        return result;
                    }

                    Debug.Assert(result != 0);
```